hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -292,10 +382,18 @@ object InferPType {
         fd.setRequired(t.required && fd.required)
       }
       case MakeTuple(values) => {
-        PTuple(values.map(v => {
+        var allRequired = false
+        val inferredVals = values.map(v => {
           InferPType(v._2, env)
+
+          if(allRequired == true && v._2.pType2.required == false) {
+            allRequired = false
+          }
+
           v._2.pType2
-        }): _*)
+        })
+
+        PTuple(allRequired, inferredVals: _*)
```

Comment:
(Similarly, this tuple is required, in that it always exists---the fields can be missing, but the tuple itself must always be present.)

Reply:
I don't quite understand, same as the MakeArray. If it's elements are not all required, then we shouldn't be able to elide the missing byte header? If I set required on the PTuple, doesn't that indicate that all of its children are also required (guaranteed non-missing)

Reply:
no, setting required on the PTuple says that the tuple itself is not allowed to be missing---the elements can do whatever they please.

Reply:
> If I set required on the PTuple, doesn't that indicate that all of its children are also required (guaranteed non-missing)

Ah, I think this must be a source of a lot of your confusion. Definitely not! Translating to scala types, the type of `PTuple(false, PInt32(true), PInt32(false))` is `Option[(Int, Option[Int])]`, while the type of `PTuple(true, PInt32(true), PInt32(false))` is `(Int, Option[Int])`. It's perfectly acceptable for an element of a tuple to be missing even if the tuple itself can't be missing!

In general, every level of a complex type can be required or not completely independently. There is no propagation to children.

Reply:
Yep, that makes sense! 

Thanks Arcturus, Tim, Patrick, I understand now.